proposed/http-factory/http-factory.md

```diff
@@ -0,0 +1,152 @@
+HTTP Factories
+==============
+
+This document describes a common standard for factories that create [PSR-7][psr7]
+compliant HTTP objects.
+
+PSR-7 did not include a recommendation on how to create HTTP objects, which leads
+to difficulty when needing to create new HTTP objects within components that are
+not tied to a specific implementation of PSR-7.
+
+The interfaces described in this document describe methods by which PSR-7 objects
+can be instantiated.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119][rfc2119].
+
+[psr7]: http://www.php-fig.org/psr/psr-7/
+[rfc2119]: http://tools.ietf.org/html/rfc2119
+
+## 1. Specification
+
+An HTTP factory is a method by which a new HTTP object, as defined by PSR-7,
+is created. HTTP factories MUST implement these interfaces for each object type
+that is implemented.
+
+## 2. Interfaces
+
+The following interfaces MAY be implemented together within a single class or
+in separate classes.
+
+### 2.1 RequestFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface RequestFactoryInterface
+{
+    /**
+     * Create a new request.
+     *
+     * @param string|UriInterface $uri
+     * @param string $method
+     *
+     * @return RequestInterface
+     */
+    public function createRequest($uri = null, $method = null);
+}
+```
+
+### 2.2 ResponseFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface ResponseFactoryInterface
+{
+    /**
+     * Create a new response.
+     *
+     * @param string|StreamInterface $body
+     *
+     * @return ResponseInterface
+     */
+    public function createResponse($body = null);
+}
+```
+
+### 2.3 ServerRequestFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface ServerRequestFactoryInterface
+{
+    /**
+     * Create a new server request.
+     *
+     * @param string|UriInterface $uri
+     * @param string $method
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequest($uri = null, $method = null);
+}
+```
+
+### 2.4 StreamFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface StreamFactoryInterface
+{
+    /**
+     * Create a new stream.
+     *
+     * @param string|resource $body
```

Comment:
I think we did this for convenience:

https://github.com/php-http/message/blob/master/src/MessageFactory/DiactorosMessageFactory.php#L40

However I think that's a design flaw, because as @shadowhand point: it's not a new stream, BUT it does rewind the stream:

https://github.com/php-http/message/blob/master/src/StreamFactory/DiactorosStreamFactory.php#L35

This is probably not a good idea either, but we faced the problem that during stream creation, we have to actually rewind it.